docs/architecture/adr-001-autopilot-oci-basic-auth-support.md

```diff
@@ -0,0 +1,137 @@
+}
+```
+
+The secret pointed by the provided `ArtifactPullSecret` will be used for pulling artifacts using either HTTP[S] or OCI protocols and is expected to by of type `kubernetes.io/dockerconfigjson` if the protocol in use is `oci://` or of type `Opaque` if protocols `http://` or `https://` are used (see below for details on the Secret layout).
+
+Example configuration for OCI:
+
+```yaml
+url: oci://my.registry/binaries/k0s:v1.30.1+k0s.0
+sha256: e95603f167cce6e3cffef5594ef06785b3c1c00d3e27d8e4fc33824fe6c38a99
+artifactPullSecret:
+  namespace: kube-system
+  name: artifacts-registry
+```
+
+Example configuration for HTTPS:
+
+```yaml
+url: https://my.file.server/binaries/k0s-v1.30.1+k0s.0
+sha256: e95603f167cce6e3cffef5594ef06785b3c1c00d3e27d8e4fc33824fe6c38a99
+artifactPullSecret:
+  namespace: kube-system
+  name: artifacts-basic-auth
+```
+
+Example configuration for HTTP:
+
+```yaml
+url: http://my.file.server/binaries/k0s-v1.30.1+k0s.0
+sha256: e95603f167cce6e3cffef5594ef06785b3c1c00d3e27d8e4fc33824fe6c38a99
+artifactPullSecret:
+  namespace: kube-system
+  name: artifacts-token-based-auth
+```
+
+### Secrets Layout
+
+For secrets of type `kubernetes.io/dockerconfigjson` the format is the default for Docker authentications, equal to what is used in a Pod's pull secret. For further details you can refer to the [official documentation](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/).
+
+When it comes to the `Opaque` secret layout (used for HTTP requests) Autopilot will accept the following entries:
+
+- `username` and `password`: if both are set then Autopilot will attempt to pull the artifacts using [Basic Authentication](https://www.ibm.com/docs/en/cics-ts/6.1?topic=concepts-http-basic-authentication).
+- `authorization`: if this property is set then the `Authorization` [header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization) will be set to its value when pulling the artifacts.
+
+No other property will be parsed and used. For sake of defining the expected behaviour in case of conflicting configurations:
+
+> In the case where the three properties are set (`username`, `password`, and `authorization`) Autopilot will ignore `username` and `password`, i.e. `authorization`  takes precedence over username and password.
+
+The `authorization` entry is used as is, its content is placed directly into the `Authorization` header. For example a secret like the following will make Autopilot to set the `Authorization` header to `Bearer abc123def456ghi789jkl0`:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: creds
+  namespace: kube-system
+data:
+  authorization: "Bearer abc123def456ghi789jkl0"
+```
+
+### Additional Details
+
+- The `Insecure` property is equivalent of defining `InsecureSkipVerify` on a Go HTTP client.
+- The `Insecure` property will be valid for both `oci://` and `https://` protocols.
+- If no protocol is defined, HTTP is used.
```

Comment:
Defaulting to a non secure protocol is quite concerning for a few reasons:
1- Since the SHA256 is optional, so if a user doesn't set either they're subject to MITM attacks delivering malicious images 
2- The authentication will be sent plain

We should probably error if the protocol isn't explicitly defined, or at least default to a secure protocol...

Reply:
I am afraid this will make the change not backward compatible.

Reply:
Interesting. So a `http.Client` assumes a "http" scheme when there's no scheme in the URL? How would such an URL look like?

Reply:
I gonna circle back on this. IIRC we use the `http.Client` by means of `grab`.

Reply:
I was wrong, `grab` simply fails if no protocol is present in the URL. I have made the default to `https` instead. Thanks for bringing this up @juanluisvaladas .

```
unsupported protocol scheme ""
```

Reply:
For sake of completeness, this seems to be the go http client's behaviour. TIL.

Reply:
I'm still in favour of requiring an explicit scheme in the URL, which we should verify and fail if no or an unknown scheme is used.

Reply:
Sounds good to me. Adjusted.